src/lib/common/sol-mainloop-impl-contiki.c

```diff
@@ -0,0 +1,287 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <contiki.h>
+#include <lib/sensors.h>
+
+#include "sol-event-handler-contiki.h"
+#include "sol-mainloop-common.h"
+#include "sol-mainloop-contiki.h"
+#include "sol-mainloop-impl.h"
+#include "sol-vector.h"
+
+#define DEFAULT_USLEEP_TIME 10000
+#define DEFAULT_USLEEP_TIME_TICKS (CLOCK_SECOND * DEFAULT_USLEEP_TIME) / NSEC_PER_SEC
+
+static process_event_t event;
+static process_data_t event_data;
+static struct etimer et;
+
+static struct sol_ptr_vector event_handler_vector = SOL_PTR_VECTOR_INIT;
+static bool event_handling_processing;
+static unsigned int event_handler_pending_deletion;
+
+struct sol_event_handler_contiki
+{
+    process_event_t *ev;
+    process_data_t ev_data;
+    void (*cb)(process_event_t ev, process_data_t ev_data, void *user_data);
+    void *data;
+    bool delete_me;
+};
+
+static void event_dispatch(void);
+
+void
+sol_mainloop_impl_lock(void)
+{
+}
+
+void
+sol_mainloop_impl_unlock(void)
+{
+}
+
+bool
+sol_mainloop_impl_main_thread_check(void)
+{
+    return true;
+}
+
+void
+sol_mainloop_impl_main_thread_notify(void)
+{
+}
+
+int
+sol_mainloop_impl_init(void)
+{
+    sol_mainloop_common_init();
+    return 0;
+}
+
+void
+sol_mainloop_impl_shutdown(void)
+{
+    void *ptr;
+    uint16_t i;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&event_handler_vector, ptr, i) {
+        free(ptr);
+    }
+    sol_ptr_vector_clear(&event_handler_vector);
+
+    sol_mainloop_common_shutdown();
+}
+
+static inline clock_time_t
+ticks_until_next_timeout(void)
+{
+    struct sol_timeout_common *timeout;
+    struct timespec now, diff;
+
+    timeout = sol_mainloop_common_timeout_first();
+    if (!timeout)
+        return DEFAULT_USLEEP_TIME_TICKS;
+
+    now = sol_util_timespec_get_current();
+    sol_util_timespec_sub(&timeout->expire, &now, &diff);
+
+    if (diff.tv_sec < 0)
+        return 0;
+
+    return diff.tv_sec * CLOCK_SECOND +
+                (CLOCK_SECOND / NSEC_PER_SEC) * diff.tv_nsec;
+}
+
+void
+sol_mainloop_impl_run(void)
+{
+    if (!sol_mainloop_impl_main_thread_check()) {
+        SOL_ERR("sol_run() called on different thread than sol_init()");
+        return;
+    }
+
+    sol_mainloop_common_loop_set(true);
+}
+
+bool
+sol_mainloop_contiki_iter(void)
+{
+    // Another event could make process wakeup
+    etimer_stop(&et);
+
+    event_dispatch();
+
+    sol_mainloop_common_timeout_process();
+    sol_mainloop_common_idler_process();
+    sol_mainloop_common_timeout_process();
+
+    if (!sol_mainloop_common_loop_check())
+        return false;
+
+    etimer_set(&et, ticks_until_next_timeout());
+    return true;
+}
+
+void
+sol_mainloop_impl_iter(void)
+{
+    // empty
+}
+
+void *
+sol_mainloop_impl_fd_add(int fd, unsigned int flags, bool (*cb)(void *data, int fd, unsigned int active_flags), const void *data)
+{
+    SOL_CRI("Unsupported");
```

Comment:
I don't think you should implement these "Unsupported". Like we do for other stuff, do not even define it in the header (#ifdef it) so it fails to compile. Nobody should be using fd_add in platforms that do not support it.